### PR TITLE
InitManager: TargetImpl::ready() behavior change

### PR DESCRIPTION
**Pull request by silentdai**
Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Description:
TargetImpl::ready() called before Init::Manager switch the target state to `ready`.
This behavior change allows asynchronous TargetImpl::ready() invocation.
This PR remains the behavior that `destroyed TargetImpl is considered as ready`.

Risk Level: Mid

Testing: Existing tests and added unit test for InitManager

Docs Changes: Inline comment

Release Notes: 
This is a developer behavior change but not a user behavior change. Not sure if release notes is needed.

This behavior is discussed in #6904

### CODE REVIEW

**Comment by alyssawilk** [comment]
@mergeconflict would you be up for doing first pass, and tagging Harvey/Matt when you've LGTMd?

File: source/common/init/manager_impl.cc [hunk 1]
```diff
@@ -16,16 +16,26 @@ void ManagerImpl::add(const Target& target) {
   TargetHandlePtr target_handle(target.createHandle(name_));
   switch (state_) {
   case State::Uninitialized:
-    // If the manager isn't initialized yet, save the target handle to be initialized later.
-    ENVOY_LOG(debug, "added {} to {}", target.name(), name_);
-    target_handles_.push_back(std::move(target_handle));
+    if (target_handle == nullptr) {
```
Comment (mergeconflict) [inline source/common/init/manager_impl.cc]: I'd like to handle this a bit differently: I don't think `createHandle` should return `nullptr`, it should return always return a valid handle. That keeps the contract between the `Target` interface and the `ManagerImpl` easiest to reason about, and keeps logic here as simple as possible.

(Imagine, for example, you had decided instead to create a new implementation of `Target` rather than modifying the existing implementation. In that case, you'd probably try to avoid changing `ManagerImpl`).

My suggestion instead is: if the handle points to a target that's already ready, that's fine. When we call `initialize` later, the target should do the right thing (specifically, if it's already ready, it should immediately call back).

File: source/common/init/manager_impl.cc [hunk 2]
```diff
@@ -51,8 +61,9 @@ void ManagerImpl::initialize(const Watcher& watcher) {
 
     // Attempt to initialize each target. If a target is unavailable, treat it as though it
     // completed immediately.
-    for (const auto& target_handle : target_handles_) {
-      if (!target_handle->initialize(watcher_)) {
+    for (auto iter = target_handles_.begin(); iter != target_handles_.end();
```
Comment (mergeconflict) [inline source/common/init/manager_impl.cc]: I was told before not to do this (although I honestly don't remember why). In any case, I'd say that the simplest and safest way to clear the list of target handles is to just call `target_handles_.clear()` after traversing it.

File: source/common/init/target_impl.cc [hunk 3]
```diff
@@ -27,19 +27,27 @@ TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
...
+          ready();
+        }
       })) {}
 
 TargetImpl::~TargetImpl() { ENVOY_LOG(debug, "{} destroyed", name_); }
 
 absl::string_view TargetImpl::name() const { return name_; }
 
 TargetHandlePtr TargetImpl::createHandle(absl::string_view handle_name) const {
+  if (is_ready_) {
```
Comment (mergeconflict) [inline source/common/init/target_impl.cc]: See note above in `ManagerImpl`. I think `createHandle` should always return a valid handle.

File: test/common/init/manager_impl_test.cc [hunk 4]
```diff
@@ -37,6 +37,41 @@ TEST(InitManagerImplTest, AddImmediateTargetsWhenUninitialized) {
...
+  m.add(t1);
+
+  ExpectableWatcherImpl w;
+
+  w.expectReady();
+  m.initialize(w);
+  expectInitialized(m);
+}
+
+TEST(InitManagerImplTest, AddTargetAndMarkReadyBeforeIntialization) {
```
Comment (mergeconflict) [inline test/common/init/manager_impl_test.cc]: Typo: "Intialization"

File: test/common/init/manager_impl_test.cc [hunk 5]
```diff
@@ -37,6 +37,41 @@ TEST(InitManagerImplTest, AddImmediateTargetsWhenUninitialized) {
   expectInitialized(m);
 }
 
+TEST(InitManagerImplTest, AddReadyTarget) {
```
Comment (mergeconflict) [inline test/common/init/manager_impl_test.cc]: I think these tests look good, thanks!

File: source/common/init/target_impl.cc [hunk 6]
```diff
@@ -27,19 +27,27 @@ TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
       fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
```
Comment (mergeconflict) [inline source/common/init/target_impl.cc]: This is the key spot I think you need to modify. I think it needs to look something like:

```c++
if (is_ready_) {
  // If we were already ready before being called by the init manager, just signal immediately...
  watcher_handle->ready();
} else {
  // ... Otherwise, hang onto the init manager's watcher handle and initialize.
  watcher_handle_ = std::move(watcher_handle);
  fn();
}
```

File: source/common/init/target_impl.cc [hunk 7]
```diff
@@ -27,19 +27,27 @@ TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
...
+    // Init::Manager should handle a nullptr return value as the target is ready.
+    return nullptr;
+  }
   // Note: can't use std::make_unique here because TargetHandleImpl ctor is private.
   return std::unique_ptr<TargetHandle>(
       new TargetHandleImpl(handle_name, name_, std::weak_ptr<InternalInitalizeFn>(fn_)));
 }
 
 bool TargetImpl::ready() {
+  is_ready_ = true;
```
Comment (mergeconflict) [inline source/common/init/target_impl.cc]: Yup, this looks right.

File: source/common/init/target_impl.cc (hunk 6 above)
Comment (silentdai) [reply source/common/init/target_impl.cc]: Thanks! That is close to my plan B. Well, this is even better as I somehow considered `is_ready_` should be shared pointer. Will use your clean code :)

File: source/common/init/manager_impl.cc (hunk 2 above)
Comment (silentdai) [reply source/common/init/manager_impl.cc]: I am curious about the scenario you were told not to. My wild guess is the timing/threadcontext. Please let me know when it comes to your mind.
Meanwhile I will change to `clear()` for this PR

File: source/common/init/manager_impl.cc (hunk 1 above)
Comment (silentdai) [reply source/common/init/manager_impl.cc]: Make sense! It's a burden to allow returning a nullptr. Will update as per your suggestion.

File: source/common/init/target_impl.cc (hunk 6 above)
Comment (silentdai) [reply source/common/init/target_impl.cc]: @mergeconflict  It's arguable if the a ready target would invoke fn().

I prefer to keep the existing behavior that fn() would be invoked to . WDYT?

File: test/common/init/manager_impl_test.cc [hunk 8]
```diff
@@ -49,18 +49,21 @@ TEST(InitManagerImplTest, AddReadyTarget) {
 
   ExpectableWatcherImpl w;
 
+  t1.expectInitialize();
```
Comment (silentdai) [inline test/common/init/manager_impl_test.cc]: @mergeconflict add this not only because it is easier to coordinate but also less astonishment.

Comment (silentdai) [reply test/common/init/manager_impl_test.cc]: Reason: 
`AddTargetAndMarkReadyBeforeInitialization` does call `expectInitialize`
Prior to this PR this function is also called anyway.

File: source/common/init/target_impl.cc (hunk 6 above)
Comment (mergeconflict) [reply source/common/init/target_impl.cc]: I don't know, that seems a little sketchy to me, maybe... `fn()` is the target's initialization function, and if the target is already ready, wouldn't that mean `fn()` (or its equivalent) would have already been called before?

File: source/common/init/target_impl.cc [hunk 9]
```diff
@@ -27,6 +27,10 @@ TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
       fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
         watcher_handle_ = std::move(watcher_handle);
         fn();
+        // It is possible that fn() mutates the is_ready_.
```
Comment (mergeconflict) [inline source/common/init/target_impl.cc]: I don't think this logic is correct. The only way that the initialization `fn()` can mutate `is_ready_` is by calling `ready()`. You don't want to call it twice.

Comment (silentdai) [reply source/common/init/target_impl.cc]: The fact is that `TargetImpl::ready()` is idempotent. I see several places utilize this assumption.

Is the lambda abusing the fact? If so I can fix

Comment (mergeconflict) [reply source/common/init/target_impl.cc]: True, it's idempotent, but that's for the user's benefit. To me, it feels really weird to say:
```c++
if (is_ready_) {
  ready();
}
```
It also feels wrong to call `fn()`, and unnecessary to hang onto the watcher handle, if `is_ready_` is true.

File: source/common/init/target_impl.cc (hunk 6 above)
Comment (silentdai) [reply source/common/init/target_impl.cc]: For those user who call ready() only in fn(), both are fine.

For the users who may call ready() before fn(), the users will see fn() invoked prior to this PR. Since we don't have a strong opinion whether fn() should be called or not, let's not break the behavior that fn() is called in this situation.

For my use case, my fn in Target constructor is empty as the ready() depends on other signal. So I vote for NOT to assume ready() always comes from fn().

Comment (mergeconflict) [reply source/common/init/target_impl.cc]: Whoa... this doesn't make sense to me. If the initialization function for your target is empty, and your target is always ready immediately, it seems like you have no need for `Init::Manager` at all.

Comment (silentdai) [reply source/common/init/target_impl.cc]: Sorry for the unclear. 

> The initialization function is empty 

True

>  your target is always ready immediately

Not always. I don't want to introduce the complexity here so let me simplify the case I was seeing.
The listener could provide the server scope Init::Manager or per listener Init::Manager. The former one could be initialized really late but my target wants to be started as early as possible but it doesn't guarantee be ready before Init::Manager::initialize().

File: source/common/init/target_impl.cc (hunk 9 above)
Comment (silentdai) [reply source/common/init/target_impl.cc]: fixed in the latest commit

File: source/common/init/target_impl.cc (hunk 6 above)
Comment (mergeconflict) [reply source/common/init/target_impl.cc]: Hmm, that's complicated... So, in this case, even though the init manager isn't telling the target _when_ to initialize, the init manager still needs to know that the target _has_ initialized, did I get that right?

File: source/common/init/target_impl.cc [hunk 10]
```diff
@@ -25,8 +25,13 @@ bool TargetHandleImpl::initialize(const Watcher& watcher) const {
 TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
     : name_(fmt::format("target {}", name)),
       fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
-        watcher_handle_ = std::move(watcher_handle);
-        fn();
+        if (is_ready_) {
+          fn();
```
Comment (mergeconflict) [inline source/common/init/target_impl.cc]: I still don't really understand why the call to `fn()` is needed in the case where the target is already initialized.

File: source/common/init/target_impl.h [hunk 11]
```diff
@@ -68,8 +68,9 @@ class TargetImpl : public Target, Logger::Loggable<Logger::Id::init> {
 
   /**
    * Signal to the init manager that this target has finished initializing. This is safe to call
-   * any time. Calling it before initialization begins or after initialization has already ended
-   * will have no effect.
+   * any time.
+   * Notes: Calling it before initialization begins is allowed.
```
Comment (mergeconflict) [inline source/common/init/target_impl.h]: Should probably clarify what happens when you call `ready()` before the init manager has called the target.

File: source/common/init/target_impl.cc (hunk 6 above)
Comment (silentdai) [reply source/common/init/target_impl.cc]: Exactly. Actually I think the existing CDS/LDS should follow this pattern as well: cluster should be a target of listener's init manager and cluster should be updated independently. However, CDS/LDS doesn't use this at the risk that throw the ordering to ADS or user.

File: source/common/init/target_impl.cc (hunk 10 above)
Comment (silentdai) [reply source/common/init/target_impl.cc]: IMHO it's not a must-have. I leave it in this way because that is what user will see.
Suppose we have the code and happens to be executed in below order:
```
Context A                         Context B                         Context C
ManagerImpl m;
                                 TargetImpl t("foo", fn);
                                 m.add(t);
                                 ...                  
X:                                                                      t.ready();
Y: m.initialize(w);
```

Prior to this PR: 
1. Either t.ready() happens before m.initialize(w), or the reverse, we will observe that fn will be called. 
2. w will be triggered only if t.ready() happens after m.initialize(w)

This goal of this PR is to change item 2.

Our conflict is whether 1 should be maintained. I don't have a strong preference but I feel there is no gain to change 1.

Comment (mergeconflict) [reply source/common/init/target_impl.cc]: That's a really good explanation.

My assumption has always been that `fn()` (or something called by it indirectly) calls `t.ready()`. Or, in other words, if `t.ready()` has been called, there's no need for `fn()` to be called again. You've said that in your case, `fn` is actually empty, and `ready` is called by some other code, and I still sort of think that's an abuse of the init manager, but I guess we can ignore that for now :)

Here's my thinking: we can't assume that `fn` is idempotent, and I think we should assume that if `is_ready_` is true, that means initialization is done. Like, I think of it as really being `is_initialized_`. So, to your point, it would be a change in behavior to not call `fn`, but I think it's most consistent with the design intent of the init manager (that is, I think it's really fixing a latent bug).

Comment (mattklein123) [reply source/common/init/target_impl.cc]: > but I guess we can ignore that for now :)

I haven't been fully tracking this conversation, but I don't think we should ignore this for now. Can we step back and make sure we are actually solving the right problem? However we think it's best to do that is fine with me.

Comment (silentdai) [reply source/common/init/target_impl.cc]: > we can't assume that fn is idempotent

I used to consider fn() 
1. a signal which is harmless to call more than once. 
2. may have side affect other than transit state from not-ready to ready. 
It leads to my code that InitManager should call it at least once.

I start to understand your concern:
if `is_ready` is true, it doesn't make sense that the init system calls fn().

Will update following your opinion

File: test/mocks/init/mocks.h [hunk 12]
```diff
@@ -50,6 +50,23 @@ class ExpectableTargetImpl : public TargetImpl {
...
+   * chained, for example: target.expectInitialize().Times(0);
+   */
+  ::testing::internal::TypedExpectation<void()>& expectInitialize();
+
+  /**
+   * Convenience method to provide a shorthand for expectInitialize() with mocked behavior of
+   * calling `ready` immediately.
+   */
+  ::testing::internal::TypedExpectation<void()>& expectInitializeWillCallReady();
+};
```
Comment (mergeconflict) [inline test/mocks/init/mocks.h]: Need blank line after class body.

File: test/mocks/init/mocks.h [hunk 13]
```diff
@@ -50,6 +50,23 @@ class ExpectableTargetImpl : public TargetImpl {
   ::testing::internal::TypedExpectation<void()>& expectInitializeWillCallReady();
 };
 
+class ExpectableEagerTargetImpl : public EagerTargetImpl {
```
Comment (mergeconflict) [inline test/mocks/init/mocks.h]: Can the code duplication here be refactored out?

File: test/common/init/manager_impl_test.cc [hunk 14]
```diff
@@ -15,7 +15,14 @@ void expectUninitialized(const Manager& m) { EXPECT_EQ(Manager::State::Uninitial
 void expectInitializing(const Manager& m) { EXPECT_EQ(Manager::State::Initializing, m.state()); }
 void expectInitialized(const Manager& m) { EXPECT_EQ(Manager::State::Initialized, m.state()); }
 
-TEST(InitManagerImplTest, AddImmediateTargetsWhenUninitialized) {
+template <typename T> class InitManagerImplTest : public ::testing::Test {};
+
+using TargetTypes = ::testing::Types<ExpectableTargetImpl, ExpectableEagerTargetImpl>;
+
+TYPED_TEST_SUITE(InitManagerImplTest, TargetTypes);
```
Comment (mergeconflict) [inline test/common/init/manager_impl_test.cc]: Neat, I hadn't seen this before. I'm not sure if you can totally get away with it, though, if you follow my suggestion of removing the initialization function from the eager target. I think the expectation is that the two target types should not behave the same in all cases.

File: source/common/init/target_impl.cc [hunk 15]
```diff
@@ -22,23 +22,25 @@ bool TargetHandleImpl::initialize(const Watcher& watcher) const {
...
 }
 
+TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
+    : AbstractTarget(
+          name, std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
+            watcher_handle_ = std::move(watcher_handle);
+            fn();
+          })) {}
+
+TargetImpl::~TargetImpl() { ENVOY_LOG(debug, "{} destroyed", name_); }
```
[the remark this answers is not in the document]
Comment (silentdai) [reply source/common/init/target_impl.cc]: great catch! will fix

File: source/common/init/target_impl.h [hunk 16]
```diff
@@ -47,11 +47,32 @@ class TargetHandleImpl : public TargetHandle, Logger::Loggable<Logger::Id::init>
   const std::weak_ptr<InternalInitalizeFn> fn_;
 };
 
+class AbstractTarget : public Target, protected Logger::Loggable<Logger::Id::init> {
```
[the remark this answers is not in the document]
Comment (silentdai) [reply source/common/init/target_impl.h]: I see. Will fix

File: source/common/init/target_impl.h [hunk 17]
```diff
@@ -62,27 +83,42 @@ class TargetImpl : public Target, Logger::Loggable<Logger::Id::init> {
...
+ */
+class EagerTargetImpl : public AbstractTarget {
+public:
+  /**
+   * @param name a human-readable target name, for logging / debugging
+   * @fn a callback function to invoke when `initialize` is called on the handle. Note that this
+   *     doesn't take a WatcherHandlePtr (like TargetFn does). Managing the watcher handle is done
+   *     internally to simplify usage.
+   */
+  EagerTargetImpl(absl::string_view name, InitializeFn fn);
```
[the remark this answers is not in the document]
Comment (silentdai) [reply source/common/init/target_impl.h]: Sorry I didn't understand. You are right, fn doesn't needed in the unique case. Will remove.

File: test/common/init/manager_impl_test.cc (hunk 14 above)
Comment (silentdai) [reply test/common/init/manager_impl_test.cc]: Right. Once the fn is removed from `EagerTargetImpl` the existing test cases don't fit. Will refactor

File: test/mocks/init/mocks.h (hunk 13 above)
Comment (silentdai) [reply test/mocks/init/mocks.h]: `ExpectableEagerTargetImpl` doesn't need `fn` and this mock is not helpful. will remove

**Comment by mattklein123** [comment]
@silentdai can we step back for a sec?

I'm a little concerned that we are not solving the right problem here. Can you summarize what problem you are trying to solve? My understanding is that you want a resource to be `ready()` before `initialize()` is called. Is that right? For my understanding, why can you not just keep track that it is ready and then just immediately call `ready()` in the context of `initialize()`?

**Comment by silentdai** [comment]
@mattklein123 Thank you for the comment!

> My understanding is that you want a resource to be ready() before initialize() is called. Is that right?

Yes!

> For my understanding, why can you not just keep track that it is ready and then just immediately call ready() in the context of initialize()?

Yes. I can. I did it in #7440 

EagerTarget beats my code in two ways 

1. coherency: `EagerTarget`  vs `a ready_flag + a TargetImpl`. It burns brain to understand the flag should be touched by Target. I don't want invent another GUARDED_BY :)

2. ownership:
There are 4 objs: Target, InitializeFn  as arg of TargetImpl ctor, ready_flag, another initialize function which could set the ready_flag. Which owns/refers which?
EagerTarget is simple: InitializeFn owns EagerTargetImpl. InitializeFn doesn't care the timing of target is added to Init::Manager. If target is gone when it should be added to Manager, it is also fine.

**Comment by mattklein123** [comment]
> Yes. I can. I did it in #7440

My personal opinion is that init manager is already complicated enough as it is, and unless we really need to make changes to support a new use case, we probably should not. IMO it's not that complicated to just call ready() in the context of initialize(). I'm pretty sure we already did that in other places, e.g., for static clusters.

However I will defer to @mergeconflict as he has been thinking about this area a lot more closely.

**Comment by mergeconflict** [comment]
> > Yes. I can. I did it in #7440
> 
> My personal opinion is that init manager is already complicated enough as it is, and unless we really need to make changes to support a new use case, we probably should not. IMO it's not that complicated to just call ready() in the context of initialize(). I'm pretty sure we already did that in other places, e.g., for static clusters.
> 
> However I will defer to @mergeconflict as he has been thinking about this area a lot more closely.

Yeah, I honestly agree with @mattklein123. If it's possible to support your use case without changes/additions to the init code, I'd prefer that, since the init manager is already tricky enough without having to reason about multiple varieties of target with different polymorphic behaviors.

**Comment by silentdai** [comment]
OK. I will adopt the idea that check ready flag in initialize function.

Sorry my initial intention is to extract the logic from xDS to `Init` namespace since this is not a configuration business logic. 

By polymorphic behaviors I also find a RAII Target which invoke ready() is also useful but I guess you don't like either.

I appreciate your comments and feedback!
